crypto/ec/ec_lib.c

```diff
@@ -1012,11 +1012,59 @@ int ec_group_simple_order_bits(const EC_GROUP *group)
     return BN_num_bits(group->order);
 }
 
+static int ec_field_inverse_mod_ord(const EC_GROUP *group, BIGNUM *r,
+                                    BIGNUM *x, BN_CTX *ctx)
+{
+    BIGNUM *e = NULL;
+    BN_CTX *new_ctx = NULL;
+    int ret = 0;
+
+    if (ctx == NULL && (ctx = new_ctx = BN_CTX_secure_new()) == NULL)
+        goto err;
+
+    BN_CTX_start(ctx);
+    if ((e = BN_CTX_get(ctx)) == NULL)
+        goto err;
+
+    /* Check if optimized inverse is implemented */
+    if (group->mont_data != NULL) {
+        /*-
+         * We want inverse in constant time, therefore we utilize the fact
+         * order must be prime and use Fermats Little Theorem instead.
+         */
+        if (!BN_set_word(e, 2))
+            goto err;
+        if (!BN_sub(e, group->order, e))
+            goto err;
+        /*-
+         * Exponent e is public.
+         * No need for scatter-gather or BN_FLG_CONSTTIME.
+         */
+        if (!BN_mod_exp_mont(r, x, e, group->order, ctx, group->mont_data))
+            goto err;
+        /* Inverse of zero doesn't exist. Let the fallback catch it. */
+        ret = (BN_is_zero(r)) ? 0 : 1;
+    }
+
+    /* Fallback to classic inverse */
+    if (ret == 0) {
+        if (!BN_mod_inverse(r, x, group->order, ctx))
```

Comment:
For the same reason as before, this is never reachable, no? (Modulo this function being newly-added as public API. See comment below.)

If unreachable, merely having the code in there may be problematic due to speculative execution problems. See here:
https://github.com/HACS-workshop/spectre-mitigations/blob/master/crypto_guidelines.md#1-do-not-conditionally-choose-between-constant-and-non-constant-time

Safer would be to just fail operation when `group->mont_data` is missing. That means the order is even, in which case you shouldn't be doing ECDSA.

Reply:
> For the same reason as before, this is never reachable, no?

I'm happy to make this change. I feel the disconnect is due to lack of any description of what the function is supposed to do. Distilling your comments (my interpretation):

    When group->order is even, this function returns an error.
    When group->order is otherwise composite, the correctness of the output is not guaranteed.
    When x is outside the range [1, group->order), the correctness of the output is not guaranteed.
    Otherwise, this function returns the multiplicative inverse in the range [1, group->order).

Is that accurate?